include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -0,0 +1,197 @@
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <ginkgo/core/base/dense_cache.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/base.hpp>
+#include <ginkgo/core/distributed/partition.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+namespace gko {
+namespace distributed {
+
+
+template <typename ValueType>
+class Vector;
+
+
+template <typename ValueType = double, typename LocalIndexType = int32,
+          typename GlobalIndexType = int64>
+class Matrix
+    : public EnableLinOp<Matrix<ValueType, LocalIndexType, GlobalIndexType>>,
+      public EnableCreateMethod<
+          Matrix<ValueType, LocalIndexType, GlobalIndexType>>,
+      public ConvertibleTo<
+          Matrix<next_precision<ValueType>, LocalIndexType, GlobalIndexType>>,
+      public DistributedBase {
+    friend class EnableCreateMethod<
+        Matrix<ValueType, LocalIndexType, GlobalIndexType>>;
+    friend class EnablePolymorphicObject<
+        Matrix<ValueType, LocalIndexType, GlobalIndexType>, LinOp>;
+    friend class Matrix<next_precision<ValueType>, LocalIndexType,
+                        GlobalIndexType>;
+
+public:
+    using value_type = ValueType;
+    using index_type = GlobalIndexType;
+    using local_index_type = LocalIndexType;
+    using global_index_type = GlobalIndexType;
+    using global_vector_type = gko::distributed::Vector<ValueType>;
+    using local_vector_type = typename global_vector_type::local_vector_type;
+    using local_matrix_type = gko::matrix::Csr<value_type, local_index_type>;
+
+    using EnableLinOp<Matrix>::convert_to;
+    using EnableLinOp<Matrix>::move_to;
+
+    void convert_to(Matrix<next_precision<value_type>, local_index_type,
+                           global_index_type>* result) const override;
+
+    void move_to(Matrix<next_precision<value_type>, local_index_type,
+                        global_index_type>* result) override;
+
+    /**
+     * Reads a square matrix from the matrix_data structure and a global row
```

Comment:
The matrix storage only works under the assumption that input and output vector have the same partition, otherwise you need to use two different partition objects for row and column indices. This should *theoretically* be possible to handle, but requires quite a few modifications and a separate read_distributed(data, row_partition, col_partition), so I think we should not add that yet. But based on that, maybe `diag` and `offdiag` should instead be named something like `local` and `non_local` (there are probably better names for that :D)

Reply:
TBH, I think adding that will be straight forward. The changes to read_distributed and its kernel are relatively minor, so I will add it in this PR.